rfcs/geometric-primitives.md

```diff
@@ -0,0 +1,419 @@
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
+struct Angle(f32);
+```
+### Traits
+
+```rust
+/// A trait for types that can be used to generate a [Mesh].
+trait Meshable{
+  fn mesh(&self, subdivisions: u8) -> Mesh;
+};
+
+/// Convex geometric primitives that are fully defined in 3D space, and can be used to bound entities within their volume. [BoundingVolume]s are useful for raycasting, collision detection, physics, and building BVHs.
+trait BoundingVolume {
+  fn raycast(&self, ray: Ray) -> Option(Point);
+}
+
+/// Convex geometric primitives that are fully defined in 2D space, and can be used to bound entities within their area. Similar to [BoundingVolumes] but for 2D instead of 3D space.
+trait BoundingArea {
+  fn raycast(&self, ray: Ray2d) -> Option(Point2d);
+}
+```
+
+### 3D Geometry Types
+
+```rust
+/// Point in 3D space
+struct Point(Vec3)
+
+/// Vector direction in 3D space that is guaranteed to be normalized through its getter/setter.
+struct Direction(Vec3)
+
+// Plane in 3d space defined by a point on the plane, and the normal direction of the plane
+struct Plane {
+  point: Point,
+  normal: Direction,
+}
+impl Meshable for Plane {}
+
+/// Unbounded line in 3D space with directionality
+struct Axis { 
+  point: Point, 
+  normal: Direction,
+}
+
+/// A newtype that differentiates an axis from a ray, where an axis in infinite and a ray is directional half-line, although their underlying representation is the same.
+struct Ray(Axis);
+
+/// A line segment bounded by two points
+struct Line { 
+  start: Point, 
+  end: Point,
+}
+impl Meshable for Line {}
+
+/// A line drawn along a path of points
+struct PolyLine {
+  points: Vec<Point>,
+}
+impl Meshable for PolyLine {}
+
+struct Triangle([Point; 3]);
+impl Meshable for Triangle {}
+
+struct Quad([Point; 4]);
+impl Meshable for Quad {}
+
+/// A sphere centered on the origin
+struct Sphere{
+  radius: f32,
+}
+impl Meshable for Sphere {}
+
+/// A cylinder with its origin at the center of the volume
+struct Cylinder {
+  radius: f32,
+  height: f32,
+}
+impl Meshable for Cylinder {}
+
+/// A cone with the origin located at the center of the circular base
+struct Cone{
+  /// Origin of the cone located at the center of its base.
+  origin: Point,
+  /// The extent of the cone is the vector that extends from the origin to the tip of the cone.
+  extent: Vec3,
+  base_radius: f32,
```

Comment:
Should this be renamed in order to be more consistent with what we see for cylinder and sphere?
```suggestion
  height: Vec3,
  radius: f32,
```